Receipts 3 - printing

malinajirka opened the pull request:
Parent issue #4039 

This PR adds support for printing receipts using Android Printing Framework.

The flow is following:
- PaymentSucceeds and ReceiptPaymentInfo is returned from CardReaderManager
- The app shows Payment successful screen with a print button
- User Clicks on the print button
- ViewModel creates ReceiptData using ReceiptDataMapper (unfortunately, ReceiptDataMapper needs to live in the app not in CardReaderModule as it uses WCOrderModel)
- ViewModel creates html receipt using ReceiptCreator
- The html receipt is passed to the view layer (CardReaderPaymentDialog)
- The view layer passes the receipt to a hidden webview which automatically starts the android printing service (using PrintHtmlHelper)

Notes: 
1. The VM also supports "SendReceipt" event, but it's not implemented in the view layer yet (I encountered some issues)
2. Tests for this PR were added in a separate PR to keep the scope of the PRs smaller

To Test:
- Make sure WCPay is installed and activated in test mode on your testing site
1. Open a detail of an unpaid order in USD
2. Tap on "Collect payment" button
3. Go through the connect and payment flows
4. Click on "Print receipt" button
5. Notice a system UI with the receipt is shown
- You can perform other tests to verify that the receipts contains - list of products, discounts, fees, shipping, taxes and total

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

malinajirka committed:
Change total amount from float to BigDecimal

malinajirka committed:
Naive implementation of receipt printing

malinajirka committed:
Extract print related code to PrintHtmlHelper.kt

malinajirka committed:
Pass document name to PrintService

malinajirka committed:
Move ReceiptLineItem creation to separate class

malinajirka committed:
Add todo

malinajirka committed:
Replace hardcoded strings with resources

malinajirka committed:
Move ReceiptData mapping to standalone class

malinajirka committed:
Fix tests

malinajirka committed:
Add tests for CardReaderPaymentViewModelTest.kt

malinajirka committed:
Init ReceiptDataMapperTest tests

malinajirka committed:
Merge branch 'develop' into issue/4039-receipts-3

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt

malinajirka committed:
Fix injection of ReceiptCreator.kt

malinajirka committed:
Add foo test

malinajirka committed:
Remove unused imports

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/PrintHtmlHelper.kt:
```diff
@@ -0,0 +1,46 @@
+
+class PrintHtmlHelper @Inject constructor() {
+    // Hold an instance of the WebView object so it isn't garbage collected before the print job is created
+    private var webViewInstance: WebView? = null
```
If I'm not mistaken `PrintHtmlHelper` can be application scoped (singleton) and you can pass application context to its constructor and it's can be used instead of an activity. In this case, you don't need to store this instance here and no need to check if it's null or not. Just local variable `webViewInstance` should do just fine

kidinov replied:
Not tested example:

```
@Singleton
class PrintHtmlHelper @Inject constructor(private val application: Application) {
    fun printReceipt(htmlReceipt: String, documentName: String) {
        val webView = WebView(application)
        webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest) = false

            override fun onPageFinished(view: WebView, url: String) {
                enqueuePrintJob(view, documentName)
            }
        }

        webView.loadDataWithBaseURL(null, htmlReceipt, "text/HTML", "UTF-8", null)
    }

    private fun enqueuePrintJob(webView: WebView, documentName: String) {
        (application.getSystemService(Context.PRINT_SERVICE) as? PrintManager)?.print(
            documentName,
            webView.createPrintDocumentAdapter(documentName),
            PrintAttributes.Builder().build()
        )
    }
}

```

malinajirka replied:
I'm not sure I follow your suggestion, so let me try to clarify my original reasoning.

After we invoke `webView.loadDataWithBaseURL(...)` it can take seconds before `onPageFinished` is invoked. During this time no-one is keeping a reference to the WebView and it can be garbage collected. If we turn PrintHtmlHelper into a singleton I don't think it'll help with this issue in any way, or am I missing something?

kidinov replied:
You make it "garbage collectible" when you put it into a property. What I'm saying is that I don't see the reason for doing that. The code looks just more complicated. If I'm not missing something the code I posted above should work just fine.

Another point was that you use activity in order to have context and pass it as a parameter (creating a potential leak, extra parameter, lifecycle issues etc) if application context should do just fine.

If you are worried about the situation when 2 or more `printReceipt` are running you could just close the dialog after print receipt button is clicked

malinajirka replied:
**Garbage collection**
Imho, all objects are garbage collectable, no matter whether they are stored in a local variable or a class-level field. The difference is that when it's just a local variable, the reference to the object is stored on stack, however, when it's a field both the object itself and the reference to that object is on the heap. Please correct me if I'm wrong, I haven't read about this topic since university so there is a pretty high chance I forgot some important details.

Now, about the code.
1. The `printReceipt` method finishes, right after `loadDataWithBaseURL` finishes. This method uses a callback to inform the caller about the progress of the loading.
2. When `printReceipt` finishes, the webview becomes unreachable from the root of our reference tree -> this means that it's eligible for garbage collection. So if we don't store the webview into a field, the garbage collector might destroy the webview before `onPageFinished` gets invoked.
3. To prevent this from happening, we need to store the webview into a field -> if we do that, our reference tree has a reference to CardReaderPaymentDialog -> which has a reference to PrintHtmlHelper -> which has a reference to the webview, so the webview is NOT eligible for garbage collection.

This is actually mentioned even in the official documentation [here](https://developer.android.com/training/printing/html-docs#load-html). 

Let me know if my reasoning makes more sense now or if you can see some gaps in it :).

**Using application instead of an activity**
I've tried to use application context instead of an activity and it's not allowed by the system - throws IllegalStateException.

> Another point was that you use activity in order to have context and pass it as a parameter (creating a potential leak, extra parameter, lifecycle issues etc) if application context should do just fine.

I believe that as long as the PrintHtmlHelper is saved just in an activity, we can't create a leak of that activity. Tbh I have never created a view using application context, I'm not sure if it's not allowed in general or the IllegalStateException is thrown just in this printing scenario. Either way, to me, having a view with application context feels more risky when it comes to leaks than using an activity. Wdyt?

kidinov replied:
> Imho, all objects are garbage collectable, no matter whether they are stored in a local variable or a class-level field. The difference is that when it's just a local variable, the reference to the object is stored on stack, however, when it's a field both the object itself and the reference to that object is on the heap

You 100% right here. It's just if there is a reference to the object in the heap in the stack then it can not be GC, otherwise, we could end up just in some weird things like reference returns a wrong object or something like that =)

> This is actually mentioned even in the official documentation here.

oh, I finally got it. The stack is cleared before `onPageFinished` is called, because (obviously) another thread is used to load the page. Although, I still can not believe that `onPageFinished` can return null instead of `webView` as a parameter because it's somehow was GC? It would be really weird behavior because it means that somehow page was rendered in the webview which not even on the heap? Maybe then `onPageFinished` is never called then but I tried to google if this happens, but could not find anything. Still confused about that


> I've tried to use application context instead of activity and it's not allowed by the system - throws IllegalStateException.

I was totally confused by the documentation:
```
    /**
     * Construct a new WebView with a Context object.
     * @param context A Context object used to access application assets.
     */
    public WebView(Context context) {
        this(context, null);
    }
```

They ask `context` not `activity` here. Also, they say that it's need to `access application assets` and you can access them via `application` context.

In any case. Thanks for helping me understand. Let's merge it!

malinajirka replied:
>  It's just if there is a reference to the object in the heap in the stack then it can not be GC, otherwise, we could end up just in some weird things like reference returns a wrong object or something like that =)

I agree that if a reference is on the stack, that garbage collector won't collect it. However, if a reference is stored on the heap, but the object is accessible from the root of the reference tree, garbage collector won't collect it either (unless it's a weak/soft reference of course). Anyway, I think we are on the same page ;).

> I still can not believe that onPageFinished can return null instead of webView as a parameter because it's somehow was GC?

`Maybe then onPageFinished is never called then` this is 100% correct. When the webview gets garbage collected, both the webview and the callback get garbage collected -> the method never gets invoked.

> They ask context not activity here. Also, they say that it's need to access application assets and you can access them via application context.

Interesting, you are correct. Webview can be apparently created with both activity and application contexts. It seems the `print` method is what is causing the IllegalStateException
```
java.lang.IllegalStateException: Can print only from an activity
        at android.print.PrintManager.print(PrintManager.java:518)
```

kidinov commented on libs/cardreader/src/main/java/com/woocommerce/android/cardreader/receipts/ReceiptCreator.kt:
```diff
@@ -131,6 +133,12 @@ class ReceiptCreator {
+     * Shows decimal places only when the number is not an integer
+     */
+    private fun formatFloat(number: Float): String =
+        if (number.rem(1f).equals(0f)) number.toInt().toString() else "%.2f".format(number)
```
Maybe something like that would be better for readability:

```
    private fun formatFloat(number: Float): String =
        if (number.isWhole()) number.toInt().toString() else "%.2f".format(number)

    private fun Float.isWhole() = rem(1f) == 0f
```

I personally couldn't grasp quickly without the comment this line `number.rem(1f).equals(0f)`.

malinajirka replied:
💯 good improvement!

I've used "isInteger" instead of "isWhole" since whole numbers can't be negative and technically, the function returns true for negative integer numbers.

Fixed in 2c8a4b0c0ec251b480c608108a06089c8027a5b8

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/PrintHtmlHelper.kt:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private fun enqueuePrintJob(activity: Activity, webView: WebView, documentName: String) {
+        (activity.getSystemService(Context.PRINT_SERVICE) as? PrintManager)?.print(
```
is this a valid case that `PRINT_SERVICE` is not available in the system? If yes, shouldn't we show something to a user? Currently it will silently do nothing if I got it correct

malinajirka replied:
I believe it's not a valid use case on Android 4+. Would you prefer if the app crashed in this scenario?

kidinov replied:
I would crash because we don't support android 4 anyway

malinajirka replied:
Updated in 5e76dc6b1c2055294ecb0b705f0a436d5fc0dcec

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:
```diff
@@ -140,9 +160,38 @@ class CardReaderPaymentViewModel @Inject constructor(
+    }
+
+    private suspend fun buildHtmlReceipt(receiptPaymentInfo: ReceiptPaymentInfo): String? {
+        return loadOrderFromDB()?.let { order ->
```
I think if the order is no in the DB at this point then something went wrong? Shouldn't we either show something to the user or at least log this situation?

malinajirka replied:
Yes, logging this situation sounds good ;). Fixed in a3d9bf9b2921993cd868bde0ca6a24d640c0d3c4

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/receipts/ReceiptDataMapper.kt:
```diff
@@ -0,0 +1,93 @@
+        }
+
+    private fun getDiscounts(order: WCOrderModel) =
+        order.discountTotal.toBigDecimalOrNull()?.roundError()?.let { total ->
```
`discountTotal` can be not a number? Shouldn't we log it somehow then? Same goes to other numbers in this file

malinajirka replied:
I agree with you that it's confusing and I don't have enough context (I don't know the history). Tbh I just wanted to be consistent with the other files in the codebase. I think logging an exception at every location where we access total/discount/fee/subtotal/tax/.... throughout the codebase is an overkill. Ideally, we'd be catching these odd cases during deserialization from the API. But that's above the scope of this PR. Wdyt?

kidinov replied:
yeah makes sense

kidinov reviewed (approved):
The code looks good and works as well. Just a couple of small things you might consider as an improvement.  Please ignore them if don't think that they are actually an improvement =)

malinajirka committed:
Minor refactoring to improve readability

malinajirka committed:
Add log when order not found

malinajirka commented:
Thanks for the review @kidinov! I've replied to your comments, let me know what you think.

malinajirka committed:
Crash when print manager not available

The default print manager should be available on all devices running Android 4+
